source/common/config/subscription_factory_impl.cc

```diff
@@ -51,6 +41,24 @@ SubscriptionPtr SubscriptionFactoryImpl::subscriptionFromConfigSource(
     const envoy::config::core::v3::ApiConfigSource& api_config_source = config.api_config_source();
     Utility::checkApiConfigSourceSubscriptionBackingCluster(cm_.primaryClusters(),
                                                             api_config_source);
+    const auto transport_api_version = api_config_source.transport_api_version();
+    if (transport_api_version == envoy::config::core::v3::ApiVersion::AUTO ||
```

Comment:
Do we want to remove AUTO along with v2? If not, shall AUTO imply v3, or via runtime to imply v3?

Reply:
I think AUTO semantics should indicate "the lowest supported version in the Envoy binary". This means that during the v2 deprecation process, it needs to fatal-by-default. When we move to V3, AUTO will then imply V3. What this means is that AUTO is kind of dangerous to use around major version transitions, but I don't see any other sensible semantics. We can't just flip AUTO to V3 without breaking people.

Reply:
Hmm yeah. Seems sub-optimal we are going to make people have to specify this. Yet more blowback will flow, but I guess we just need to get this over with.

Reply:
> When we move to V3, AUTO will then imply V3. What this means is that AUTO is kind of dangerous to use around major version transitions, but I don't see any other sensible semantics. We can't just flip AUTO to V3 without breaking people.

Right, so my point was adding another runtime flag (disabled by default), that makes AUTO as V3 instead of failing here. And later we can make it enabled by default? So people can either specifying V3 everywhere or enable the flag to accommodate V3 upgrade? If users will need to specify V3 explicitly everywhere to upgrade, then what's the point of having AUTO?

Reply:
There are three options I can see here:

1. Fundamentally AUTO is confusing and we should get rid of it, so I'm down for removing it along with v2.

2. Having to explicitly specify version means that we need to have more boiler plate and make everything _transport_ major version aware. We already have to do that for `resource_api_version`, so this is not such a big deal? If we do want to remove this additional boilerplate we should follow the approach described in https://github.com/envoyproxy/envoy/pull/14223#discussion_r533775461.

3. We could consider mapping AUTO to automatic discovery of the transport endpoint by probing. I'm not a huge fan of this as it is potentially fragile and might be confusing during rollouts.

What do folks prefer? I think I would prefer 1, 2, 3 in that order, but I'm happy to go with the weight of opinion here.

Reply:
I would vote for (1). I think AUTO is too magical for something that has to be explicitly specified in practice and is therefore not actually auto in any way. :)

Reply:
Same preference in that order. I don't think I'd want 3 though.